frontend/catalyst/jax_tracer.py

```diff
@@ -1225,9 +1261,24 @@ def is_leaf(obj):
                 meas, meas_trees = trace_quantum_measurements(device, qrp_out, output, trees)
                 qreg_out = qrp_out.actualize()
 
-                meas_tracers = [trace.full_raise(m) for m in meas]
+                # Check if the measurements are nested then apply the full_raise
+                def check_full_raise(arr, func):
+                    if isinstance(arr, list):
+                        return [check_full_raise(x, func) for x in arr]
+                    else:
+                        return func(arr)
+
+                meas_tracers = check_full_raise(meas, trace.full_raise)
                 meas_results = tree_unflatten(meas_trees, meas_tracers)
 
+                # Always return a tuple of results with shot-vectors where num_of_total_copies > 1
+                if get_device_total_shot_copies(device) > 1:
```

Comment:
Minor: here I think you can handle any shape of return by using pytrees of the return values. https://jax.readthedocs.io/en/latest/pytrees.html You can then handle any return shape even something that mixes dict and list. "{"first": qml.sample(), "second": [100, qml.sample()]}".  They are multiple place in the codebase where we use that.

Reply:
Thanks for the tip! It sounds like a robust solution. I’ll look into integrating this approach. Appreciate the guidance!

Reply:
Just a note that arbitrary output containers are already being restored on the line just above:
`meas_results = tree_unflatten(meas_trees, meas_tracers)`. 

The processing you are doing below looks to me like it just replaces lists with tuples (bare and inside one layer of dictionaries). I think this is unnecessary, and is probably better handled by just replacing the list you produce in `trace_quantum_measurements` with a tuple directly in that function.

You can still add the test case Romain suggested to make sure more complex container structures will work though :)